packages/compose/src/utils/create-higher-order-component/index.ts

```diff
@@ -0,0 +1,58 @@
+/**
+ * External dependencies
+ */
+import { camelCase, upperFirst } from 'lodash';
+// eslint-disable-next-line no-restricted-imports
+import type { ComponentType } from 'react';
+import type { Subtract } from 'utility-types';
+
+/**
+ * Higher order components can cause props to be obviated. For example a HOC that
+ * injects i18n props will obviate the need for the i18n props to be passed to the component.
+ *
+ * If a HOC does not obviate the need for any specific props then we default to `{}` which
+ * essentially subtracts 0 from the original props of the passed in component. An example
+ * of this is the `pure` HOC which does not change the API surface of the component but
+ * simply modifies the internals.
+ */
+export interface HigherOrderComponent< TObviatedProps extends object = {} > {
+	< TP extends TObviatedProps >(
+		OriginalComponent: ComponentType< TP >
+	): ComponentType< Subtract< TP, TObviatedProps > >;
+}
+
+/**
+ * Given a function mapping a component to an enhanced component and modifier
+ * name, returns the enhanced component augmented with a generated displayName.
+ *
+ * @param mapComponentToEnhancedComponent Function mapping component to enhanced component.
+ * @param modifierName                    Seed name from which to generated display name.
+ *
+ * @return Component class with generated display name assigned.
+ */
+function createHigherOrderComponent<
+	TObviatedProps extends object,
+	TProps extends TObviatedProps
+>(
+	mapComponentToEnhancedComponent: (
+		OriginalComponent: ComponentType< TProps >
+	) => ComponentType< Subtract< TProps, TObviatedProps > >,
```

Comment:
Fun fact: The type of `mapComponentToEnhancedComponent` is `HigherOrderComponent<TObviatedProps>`.

Reply:
While that's true I don't think we can use the type here becuase we can't pass `TProps` down, as the `HOC` type only accepts it on the call signature and we need to grab it from above. But it is a fun fact 😁

Reply:
OK, I hoped for a little bit more than a fun fact but yes, now I see that `HigherOrderComponent` doesn't specify all the generic types we need 🙂